modules/core/src/main/scala/com/twilio/guardrail/sbt/AbstractGuardrailPlugin.scala

```diff
@@ -142,14 +144,19 @@ trait AbstractGuardrailPlugin { self: AutoPlugin =>
     }
   }
 
-  override lazy val projectSettings = Seq(
-    Keys.guardrailTasks in Compile := List.empty,
-    Keys.guardrailTasks in Test := List.empty,
-    Keys.guardrail in Compile := Tasks.guardrailTask(runner)((Keys.guardrailTasks in Compile).value, (SbtKeys.managedSourceDirectories in Compile).value.head),
-    Keys.guardrail in Test := Tasks.guardrailTask(runner)((Keys.guardrailTasks in Test).value, (SbtKeys.managedSourceDirectories in Test).value.head),
-    SbtKeys.sourceGenerators in Compile += (Keys.guardrail in Compile).taskValue,
-    SbtKeys.sourceGenerators in Test += (Keys.guardrail in Test).taskValue,
-    SbtKeys.watchSources in Compile ++= Tasks.watchSources((Keys.guardrailTasks in Compile).value),
-    SbtKeys.watchSources in Test ++= Tasks.watchSources((Keys.guardrailTasks in Test).value),
+  override lazy val projectSettings: Seq[Def.Setting[_]] = Seq(Compile, Test).flatMap(conf =>
+    Seq(
+      conf / Keys.guardrailTasks := List.empty,
+      conf / Keys.guardrail / NioKeys.fileInputs := (conf / Keys.guardrailTasks).value.flatMap(_._2.specPath).map(_.toPath.toAbsolutePath.toGlob),
+      conf / Keys.guardrail := {
+        val tasks = (conf / Keys.guardrailTasks).value
+        val sourceDir = (conf / SbtKeys.sourceManaged).value
+        val changes = (conf / Keys.guardrail).inputFileChanges
+        if (!sourceDir.exists() || changes.hasChanges) Tasks.guardrailTask(runner)(tasks, sourceDir)
+        else (sourceDir ** "*.scala" +++ sourceDir ** "*.java").get()
```

Comment:
So, my understanding is that this would end up globbing every file, not just guardrail files, that were in the `sourceDir` and returning those from this plugin, as if we were the ones who had created them.

My experience using plugins that do this is that scalac does no deduplication of input files, the result of which is conflicts from one file back to itself.

I never really looked into it too deeply though, I just have some vague idea that this is the case. Do you know about this, or have you tested this plugin with other plugins that write to `sourceManaged` (avro, scalaxb, others?)

Reply:
No, didn't really tested with other plugins. Didn't use neither avro or scalaxb before. Could you advice what case should I test?

Yes, indeed, this glob simply takes **all files inside directory recursively** which end with `.scala` or `.java`. I understand what it could be not always what we want.

To overcome this ideally guardrail should have a very fast (comparable to the default one) method, which does nothing, but only generates list of files, not generating their content itself and not creating them in filesystem. Some sort of "--dry-run" option? Is this supported? Or how hard could it be to add?

Reply:
Yeah, `--dry-run` is a really good idea, but will require some work. As a stopgap, is it possible to store the previous file list in a session variable?

As far as a way to test this, I think you could modify this plugin to just return the list of generated files twice over, so that every file gets passed to scalac twice. This will simulate what it would be like for a misconfigured plugin.